File: src/main/golang/leetcode/editor/cn/[354]Russian Doll Envelopes.go
package lc

import "sort"

//leetcode submit region begin(Prohibit modification and deletion)
func maxEnvelopes(envelopes [][]int) int {
	sort.Slice(envelopes, func(i, j int) bool {
		if envelopes[i][0] == envelopes[j][0] {
			return envelopes[i][1] > envelopes[j][1]
		} else {
			return envelopes[i][0] < envelopes[j][0]
		}
	})
	n := len(envelopes)
	top := make([]int, n)
	var piles int
	for i := 0; i < n; i++ {
		num := envelopes[i][1]
		l, r := 0, piles
		for l < r {
			mid := l + (r-l)/2
			if top[mid] < num {
				l = mid + 1
			} else {
				r = mid
			}
		}
		if l == piles {
			piles++
		}
		top[l] = num
	}
	return piles
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[354]Russian Doll Envelopes.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        envelopes.sort(key=lambda x: (x[0], -x[1]))
        height = [x[1] for x in envelopes]

        def lengthOfLIS(nums):
            piles, n = 0, len(nums)
            top = [0] * n
            for i in range(n):
                poker = nums[i]
                left, right = 0, piles
                while left < right:
                    mid = (left + right) // 2
                    if top[mid] < poker:
                        left = mid + 1
                    else:
                        right = mid
                if left == piles:
                    piles += 1
                top[left] = poker

            return piles

        return lengthOfLIS(height)
# leetcode submit region end(Prohibit modification and deletion)
